api/src/main/java/org/openmrs/api/impl/PersonServiceImpl.java

```diff
@@ -991,4 +995,272 @@ public void checkIfPersonAttributeTypesAreLocked() {
+		// we don't want to change the preferred name of the preferred patient
+		tmpName.setPreferred(false);
+		tmpName.setUuid(UUID.randomUUID().toString());
+		return tmpName;
+	}
+	
+	private void mergeAddresses(Person preferred, Person notPreferred, PersonMergeLogData mergedData)
+	        throws SerializationException {
+		// move all addresses
+		// (must be done after all calls to services above so hbm doesn't try to save things prematurely (hacky)
+		for (PersonAddress newAddress : notPreferred.getAddresses()) {
+			boolean containsAddress = false;
+			for (PersonAddress currentAddress : preferred.getAddresses()) {
+				containsAddress = currentAddress.equalsContent(newAddress);
+				if (containsAddress) {
+					break;
+				}
+			}
+			if (!containsAddress) {
+				PersonAddress tmpAddress = (PersonAddress) newAddress.clone();
+				tmpAddress.setPersonAddressId(null);
+				tmpAddress.setVoided(false);
+				tmpAddress.setVoidedBy(null);
+				tmpAddress.setVoidReason(null);
+				tmpAddress.setPreferred(false); // addresses from non-preferred patient shouldn't be marked as preferred
+				tmpAddress.setUuid(UUID.randomUUID().toString());
+				preferred.addAddress(tmpAddress);
+				mergedData.addCreatedAddress(tmpAddress.getUuid());
+				log.debug("Merging address " + newAddress.getPersonAddressId() + " to " + preferred.getPersonId());
+			}
+		}
+		
+		// copy person attributes
+		for (PersonAttribute attr : notPreferred.getAttributes()) {
+			if (!attr.getVoided()) {
+				PersonAttribute tmpAttr = attr.copy();
+				tmpAttr.setPerson(null);
+				tmpAttr.setUuid(UUID.randomUUID().toString());
+				preferred.addAttribute(tmpAttr);
+				mergedData.addCreatedAttribute(tmpAttr.getUuid());
+			}
+		}
+		
+		// move all other person info
+		mergedData.setPriorGender(preferred.getGender());
+		if (!"M".equals(preferred.getGender()) && !"F".equals(preferred.getGender())) {
+			preferred.setGender(notPreferred.getGender());
+		}
+
+		mergedData.setPriorDateOfBirth(preferred.getBirthdate());
+		mergedData.setPriorDateOfBirthEstimated(preferred.getBirthdateEstimated());
+		if (preferred.getBirthdate() == null || (preferred.getBirthdateEstimated() && !notPreferred.getBirthdateEstimated())) {
+			preferred.setBirthdate(notPreferred.getBirthdate());
+			preferred.setBirthdateEstimated(notPreferred.getBirthdateEstimated());
+		}
+		mergedData.setPriorDateOfDeathEstimated(preferred.getDeathdateEstimated());
+		if (preferred.getDeathdateEstimated() == null) {
+			preferred.setDeathdateEstimated(notPreferred.getDeathdateEstimated());
+		}
+		
+		mergedData.setPriorDateOfDeath(preferred.getDeathDate());
+		if (preferred.getDeathDate() == null) {
+			preferred.setDeathDate(notPreferred.getDeathDate());
+		}
+		
+		if (preferred.getCauseOfDeath() != null) {
+			mergedData.setPriorCauseOfDeath(preferred.getCauseOfDeath().getUuid());
+		}
+		if (preferred.getCauseOfDeath() == null) {
+			preferred.setCauseOfDeath(notPreferred.getCauseOfDeath());
```

Comment:
This also could be moved to **else** statement without doing if (preferred.getCauseOfDeath() == null)